### ISSUE TITLE
Don't understand return values of decision_function of SLIM

### ISSUE THREAD

**Issue by alexandre-termier** [scikit-mine/scikit-mine#125]
Hello,

I have some questions about the decision_function of SLIM, especially the values it returns:

- first, why return negative values to be compatible with sklearn? I have a hard time getting why a code length could be negative, this may require more explanation in the documentation if this is important
- second, I don't get the handling of values of the log vector:
- if the original value is 0, it is left untouched in _log2, so it will be 0 at the end of decision_function
- if the original value is 1, it will be changed to 0 by the log, and later decision_function will change it to -np.inf 

This does not seem consistent. Is it a bug, or is there some underlying rationale for it ?
Thanks,
Alex

**Comment by remiadon** [comment]
This is simply to provide compatibility with sklearn

You have a complete example here 
https://scikit-mine.github.io/scikit-mine/tutorials/itemsets/symbolic_classifier.html#Shortest-code-wins-!!

**Comment by hermann74** [comment]
Indeed, the negative values in the decision_function allow the operation of OneVsRestClassifier

IIn the case of k>=3 classes, OVR creates k SLIM estimators (class i vs non i) that it fits on the labeled data (X,y), ovr_slim.fit(X,y) .


In order to predict the class of an item x, ovr_slim.predict(x) computes the k values $[d_1, ..., d_k]$of the decision_function for each class and returns the one corresponding to the maximum of these values.

Now x is "close" to a class k the shorter its code c_k is and thus close to zero.
By posing $d_i = - c_i$ , $d_i$ is then the maximum as agreed.

In fact, the length c_k being in $[0; + \infty[,$ any monotonous decreasing function f on [0, + \infty [ is suitable by posing $d_i=f(c_i)$. Currently, we have taken $f(x) = exp(-0.2x)$ in order to have $f(x)$ values in $[0,1]$ corresponding to probabilities (like a sigmoid output). 
In the code 


```
def get_code_length(D):
      .....
     return codes_length()
```

```
def decision_function(self, D):
     _fonc = lambda x: np.exp(-0.2 * x)
     return _fonc(self.get_code_length(D))
```

**Comment by hermann74** [comment]
Moreover, in the binary case $k=2$, OvR does not work because it only creates a SLIM model and makes ovr_slim.predict(x) only provide a value d that it tries to compare to 0 (see https://github.com/scikit-learn/scikit-learn/blob/98cf537f5c538fdbc9d27b851cf03ce7611b8a48/sklearn/multiclass.py#L107)
As shown on the sklearn documentation, `decision_function(x)`
 corresponds to the signed distance of $x$ to a boundary F (separating hyperplane). If $d \ge 0$ , then $x$ is predicted in class C and not C if $d<0$.
Not having this boundary, SlimClassifier instantiates 2 SLIMs and computes the distance of $x$ to class C, then of $x$ to non-C, and takes the maximum.
